File: 0206 Reverse Linked List/solution.py
#! python3
from __future__ import print_function

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        while head:
            dummy.next, head, dummy.next.next = head, head.next, dummy.next
        return dummy.next
        
def createListNode(nums):
    if len(nums) == 0:
        return None
    head = ListNode(nums[0])
    p = head
    for i in nums[1:]:
        p.next = ListNode(i)
        p = p.next
    return head

if __name__ == "__main__":
    head = createListNode([1,2,3,4,5])
    head = Solution().reverseList(head)
    while head:
        if head.next:
            print (head.val,end='->')
        else:
            print(head.val)
        head = head.next
        

File: 0206 Reverse Linked List/solution.cpp
#include "iostream"
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next)
            return head;
        ListNode *p = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return p;
    }
};

ListNode *create_linkedlist(std::initializer_list<int> lst)
{
    auto iter = lst.begin();
    ListNode *head = lst.size() ? new ListNode(*iter++) : NULL;
    for (ListNode *cur=head; iter != lst.end(); cur=cur->next)
        cur->next = new ListNode(*iter++);
    return head;
}

int main(){

    ListNode *head = Solution().reverseList(create_linkedlist({1,2,3,4,5,6}));
    for (ListNode *cur = head; cur; cur = cur->next)
        cout << cur->val << "->";
    cout << "\b\b  " << endl;
    return 0;
}
